### ISSUE TITLE
Batch RPC calls to bitcoind

### ISSUE THREAD

**Issue by shesek** [shesek/bwt#5]
Not yet implemented in rust-bitcoincore-rpc: https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/27

**Comment by dpc** [comment]
From my experience batching Bitcoin request and json-rpc layer is totally not worth it. Just use a threadpool.

**Comment by shesek** [comment]
Thanks for the tip!

Did you try with high-latency connections, say over Tor?

**Comment by dpc** [comment]
> Did you try with high-latency connections, say over Tor?

No. The use cases I've worked with were all reasonably low latencies.

Just running this in my head, I don't see how batching would be better there either.

The optimization problem here is basically saturating IO throughput both on the network on disk of the fullnode. Naive implementation blocking on each rpc call under-utilizes resources, wasting the latency of the whole round-trip on each call.

Batching gets rid of most round-trips, but still suffers from the round-trip. Just once in N times instead of every time. A thread-pool can keep both the link/fullnode IO saturated 100% time. The amount of work (like serialization&deserialization) is roughly the same -  with threadpool there's just more HTTP & JSONRPC envolope parsing, but these are minuscule.

On a long latency link like Tor, one would need more threads in the threadpool to stature the IO ... assuming high throughput. But since both the throughput and latency are (I assume) going to be low, then in practice a threadpool of 2 is going to saturate the connection.
